### PR TITLE
`Communication`: Add all participants to default channels automatically

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

#6487 creates 4 default channels when a course is created, but participants are not added automatically yet.

### Description

Upon course creation, the 4 default channels are created and all students, tutors, and instructors of the course are added to them.
Additionally, if a user is added to a group (which potentially belongs to multiple courses) they join all the default channels of all the courses the user is added to.

### Steps for Testing

Prerequisites:
- 2 Instructors
- 1 Student
- 1 Tutor
- (1 Editor)

1. Log in to Artemis as one instructor
2. Navigate to Course Administration
3. Create a course. Enable messaging
4. Add the other users (1 student, 1 tutor, 1 additional instructor, (1 editor)) to the course
5. Check if the other users are automatically added to all 4 default channels
6. Create another course with the same student/tutor/instructor(/editor) groups
7. Check if all users are automatically added to the 4 default channels of the new course

#### Performance Review
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [DefaultChannelType.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4788-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.enumeration/DefaultChannelType.html) | 100% | ✅ |
| [CourseRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4788-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/CourseRepository.html) | 81% | ✅ |
| [ChannelService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4788-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ChannelService.html) | 95% | ✅ |
| [ConversationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4788-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ConversationService.html) | 96% | ✅ |
| [CourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4788-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CourseResource.html) | 94% | ✅ |
| [AdminCourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4788-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminCourseResource.html) | 100% | ✅ |
| [ChannelResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4788-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis.conversation/ChannelResource.html) | 91% | ✅ |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java [hunk 1]
```diff
@@ -52,6 +52,15 @@ public interface CourseRepository extends JpaRepository<Course, Long> {
     @Query("select distinct course from Course course where course.studentGroupName like :name")
     Course findCourseByStudentGroupName(@Param("name") String name);
 
+    @Query("select distinct course from Course course where course.instructorGroupName = :name")
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java:55]: We usually use text blocks """ """ for SQL queries and capital letters for SQL keywords. Would be good if you change it for consistency, ideally for both your new queries and the old ones you encounter.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java [hunk 2]
```diff
@@ -137,6 +141,15 @@ public Channel createChannel(Course course, Channel channel, Optional<User> crea
         return savedChannel;
     }
 
+    public Set<User> registerUsersToChannel(Boolean addAllStudents, Boolean addAllTutors, Boolean addAllInstructors, List<String> usersLoginsToRegister, Course course,
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:144]: is there are reason, the parameters are Booleans and not booleans, can they ever be null?

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 3]
```diff
@@ -106,11 +108,16 @@ public class CourseResource {
 
     private final CourseScoreCalculationService courseScoreCalculationService;
 
+    private final ConversationService conversationService;
+
+    private ChannelRepository channelRepository;
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:113]: Is there a reason why this is not final?

**Review by tobias-lippert — COMMENTED** [review]
Code looks good in general, just some minor comments

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java (hunk 3 above)
Comment (pellumbbaboci) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:113]: Actually no, `final` will be added.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java (hunk 2 above)
Comment (pellumbbaboci) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:144]: no, will change it to `boolean` 👍 .

File: src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java (hunk 1 above)
Comment (pellumbbaboci) [reply src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java:55]: Changed to """"" text blocks

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java [hunk 4]
```diff
@@ -137,6 +141,15 @@ public Channel createChannel(Course course, Channel channel, Optional<User> crea
         return savedChannel;
     }
 
+    public Set<User> registerUsersToChannel(boolean addAllStudents, boolean addAllTutors, boolean addAllInstructors, List<String> usersLoginsToRegister, Course course,
+            Channel channel) {
+        Set<User> usersToRegister = new HashSet<>();
+        usersToRegister.addAll(conversationService.findUsersInDatabase(course, addAllStudents, addAllTutors, addAllInstructors));
+        usersToRegister.addAll(conversationService.findUsersInDatabase(usersLoginsToRegister.stream().toList()));
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:148]: This is already a list. Why do you have to create a stream and a list again?

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 5]
```diff
@@ -1007,6 +1016,22 @@ public ResponseEntity<Void> addUserToCourseGroup(String userLogin, User instruct
...
+                case TEACHING_ASSISTANT -> courseRepository.findCoursesByTeachingAssistantGroupName(group);
+                case INSTRUCTOR -> courseRepository.findCoursesByInstructorGroupName(group);
+                default -> List.of();
+            };
+
+            for (Course c : courses) {
+                channelRepository.findChannelsByCourseId(c.getId()).stream().filter(channel -> channelNames.contains(channel.getName())).forEach(channel -> {
+                    conversationService.registerUsersToConversation(c, Set.of(userToAddToGroup.get()), channel, Optional.empty());
+                });
+            }
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:1033]: I'd move this into the corresponding service

File: src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ChannelResource.java [hunk 6]
```diff
@@ -314,11 +314,8 @@ public ResponseEntity<Void> registerUsersToChannel(@PathVariable Long courseId,
...
         var requestingUser = userRepository.getUserWithGroupsAndAuthorities();
         channelAuthorizationService.isAllowedToRegisterUsersToChannel(channelFromDatabase, usersLoginsToRegister, requestingUser);
-        Set<User> usersToRegister = new HashSet<>();
-        usersToRegister.addAll(conversationService.findUsersInDatabase(course, addAllStudents, addAllTutors, addAllInstructors));
-        usersToRegister.addAll(conversationService.findUsersInDatabase(usersLoginsToRegister.stream().toList()));
-        conversationService.registerUsersToConversation(course, usersToRegister, channelFromDatabase, Optional.empty());
-        usersToRegister.forEach(user -> singleUserNotificationService.notifyClientAboutConversationCreationOrDeletion(channelFromDatabase, user, requestingUser,
+        Set<User> registeredUsers = channelService.registerUsersToChannel(addAllStudents, addAllTutors, addAllInstructors, usersLoginsToRegister, course, channelFromDatabase);
+        registeredUsers.forEach(user -> singleUserNotificationService.notifyClientAboutConversationCreationOrDeletion(channelFromDatabase, user, requestingUser,
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ChannelResource.java:318]: As far as I can see this is sent for each default channel. Can we only create one notification telling the user that they were added to all default channels (maybe mentioning them in there)?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Left some comments.
Additionally there are no tests yet. Do you plan on adding them in this PR as well?

File: src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ChannelResource.java (hunk 6 above)
Comment (pellumbbaboci) [reply src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ChannelResource.java:318]: The notifications are not sent when channels are created automatically at the moment. This notification here is sent only when a user is added manually.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java (hunk 4 above)
Comment (pellumbbaboci) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:148]: It seems that we have overseen this, will remove the unnecessary conversion

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java (hunk 5 above)
Comment (pellumbbaboci) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java]: Moved to conversation service 👍

**Review by SolizerCodes — DISMISSED** [review]
Tested on ts2. LGTM

**Review by pal03377 — APPROVED** [review]
Tested in testing session on ts1 + looked at code. Works & looks good!

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts1. All scenarios worked as described

**Review by laadvo — COMMENTED** [review]
Tested on ts1 in the testing session, the users of the course get added to the defautl channels when a new course is created. The admin that created the course also gets added to the four default channels. It might be better for the admin to not get added, so that he does not get spammed with messages from many different courses.

**Review by RY997 — APPROVED** [review]
Test on ts1, works well.

**Review by JohannesStoehr — APPROVED** [review]
Comments were addressed, code looks good to me

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java [hunk 7]
```diff
@@ -157,6 +163,30 @@ public void registerUsersToConversation(Course course, Set<User> users, Conversa
...
 
+    /**
+     * Add user to default channels of courses with the same group. This is used when a user is added to a group.
+     *
+     * @param userToAddToGroup the user to be added
+     * @param group            the group of the user
+     * @param role             the role of the user
+     */
+    public void registerUserToDefaultChannels(User userToAddToGroup, String group, Role role) {
+        final List<String> channelNames = Arrays.stream(DefaultChannelType.values()).toList().stream().map(DefaultChannelType::getName).toList();
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:174]:
```suggestion
        final Set<String> channelNames = Arrays.stream(DefaultChannelType.values()).map(DefaultChannelType::getName).collect(Collectors.toSet());

```

**Review by bassner — CHANGES_REQUESTED** [review]
Added two small suggestions :)

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java [hunk 8]
```diff
@@ -157,6 +163,30 @@ public void registerUsersToConversation(Course course, Set<User> users, Conversa
...
     }
 
+    /**
+     * Add user to default channels of courses with the same group. This is used when a user is added to a group.
+     *
+     * @param userToAddToGroup the user to be added
+     * @param group            the group of the user
+     * @param role             the role of the user
+     */
+    public void registerUserToDefaultChannels(User userToAddToGroup, String group, Role role) {
```
[the remark this answers is not in the document]
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:173]: Yes, I have moved the method.

**Review by bassner — APPROVED** [review]
Thanks for the updates. LGTM!

**Pull request merged**